modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/prepare/BaseQueryContext.java

```diff
@@ -0,0 +1,291 @@
+     */
+    private BaseQueryContext(
+        FrameworkConfig cfg,
+        Context parentCtx,
+        IgniteLogger log
+    ) {
+        super(Contexts.chain(parentCtx, cfg.getContext()));
+
+        // link frameworkConfig#context() to this.
+        this.cfg = Frameworks.newConfigBuilder(cfg).context(this).build();
+
+        this.log = log;
+
+        RelDataTypeSystem typeSys = CALCITE_CONNECTION_CONFIG.typeSystem(RelDataTypeSystem.class, cfg.getTypeSystem());
+
+        typeFactory = new IgniteTypeFactory(typeSys);
+
+        qryCancel = unwrap(GridQueryCancel.class);
+
+        rexBuilder = new RexBuilder(typeFactory);
+    }
+
+    /**
+     * @return Framework config.
+     */
+    public FrameworkConfig config() {
+        return cfg;
+    }
+
+    /**
+     * @return Logger.
+     */
+    public IgniteLogger logger() {
+        return log;
+    }
+
+    /**
+     * @return Schema name.
+     */
+    public String schemaName() {
+        return schema().getName();
+    }
+
+    /**
+     * @return Schema.
+     */
+    public SchemaPlus schema() {
+        return cfg.getDefaultSchema();
+    }
+
+    /**
+     * @return Type factory.
+     */
+    public IgniteTypeFactory typeFactory() {
+        return typeFactory;
+    }
+
+    /** */
+    public RexBuilder rexBuilder() {
+        return rexBuilder;
+    }
+
+    /**
+     * @return New catalog reader.
+     */
+    public CalciteCatalogReader catalogReader() {
+        if (catalogReader != null)
+            return catalogReader;
+
+        SchemaPlus dfltSchema = schema(), rootSchema = dfltSchema;
+
+        while (rootSchema.getParentSchema() != null)
+            rootSchema = rootSchema.getParentSchema();
+
+        return catalogReader = new CalciteCatalogReader(
+            CalciteSchema.from(rootSchema),
+            CalciteSchema.from(dfltSchema).path(null),
+            typeFactory(), CALCITE_CONNECTION_CONFIG);
+    }
+
+    /**
+     * @return Query cancel.
+     */
+    public GridQueryCancel queryCancel() {
+        return qryCancel;
+    }
+
+    /**
+     * @return Context builder.
+     */
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    /** */
+    public static BaseQueryContext empty() {
+        return EMPTY_CONTEXT;
+    }
+
+    /**
+     * Planner context builder.
+     */
+    @SuppressWarnings("PublicInnerClass") 
+    public static class Builder {
+        /** */
+        private static final RelTraitDef<?>[] TRAIT_DEFS = {
+            ConventionTraitDef.INSTANCE,
+            RelCollationTraitDef.INSTANCE,
+            DistributionTraitDef.INSTANCE,
+            RewindabilityTraitDef.INSTANCE,
+            CorrelationTraitDef.INSTANCE,
+        };
+
+        /** */
+        private static final FrameworkConfig EMPTY_CONFIG =
+            Frameworks.newConfigBuilder(FRAMEWORK_CONFIG)
+                .defaultSchema(createRootSchema(false))
+                .traitDefs(TRAIT_DEFS)
```

Comment:
`traitDefs` are already set in `FRAMEWORK_CONFIG`, it's not required to set it here.